[REG-2014] Draft of "What is Regression Games" (#95)
Regression-Games/RegressionDocs: docs/what-is-regression-games.mdx

sidebar_label: 'What is Regression Games?'
---

# Regression Games

Regression Games is making it easy to create automated tests for your game.

## Highlighted Features

| Feature | Description |
| --- | --- |

> | [Plug-and-Play Setup](getting-started/installing-regression-games) | Simply install our SDK and add one GameObject to your scene to instantly begin using our tools. Custom integration code is rarely needed to get started. |

Reviewer: Consider changing the word 'install' to be more unity specific. How difficult is it to install? I don't know. But I know how easy it is to add a unity package.

Something like: 
"Simply import our SDK package and add one GameObject to your scene to instantly begin using our tools. Custom integration code is rarely needed to get started."

Author: Added!

> | [Get Started Quickly with Recording and Input Playback](getting-started/creating-your-first-automated-test) | Developers can create functional tests by recording a play-through of their scenario and preforming validations against automated playback. |

Reviewer: Consider consistently speaking to the user directly. 

"Create functional tests by recording a play-through of your scenario and preforming validations against automated playback."

Author: Added!

> | [Computer Vision Capabilities](guides/interacting-with-ui-using-computer-vision) | In some cases, a bot needs to see the game to understand it. Our CV capabilities, such as OCR and object detection, can extract information from in-game screenshots to help drive behaviors and validate game state during runtime. |

Reviewer: Consider italicizing 'see'. Why? I don't know, it feels better.

Also consider speaking to the reader by changing 'the game' to 'your game'

"In some cases, a bot needs to _see_ your game to understand it. Our CV capabilities, such as OCR and object detection, can extract information from in-game screenshots to help drive behaviors and validate game state during runtime."

Author: Added!

> | [Exploratory Testing](guides/finding-game-breaking-bugs-with-stability-testing) | Run more than just predefined scenarios - exploratory bots such as the Monkey Bot can explore edge cases within your game. |

Reviewer 2: ```suggestion
| [Exploratory Testing](guides/finding-game-breaking-bugs-with-stability-testing) | Run more than just predefined scenarios - tools such as our Monkey Bot can dynamically explore your game to find edge cases. |
```

> | [Composability](core-concepts/bot-sequences/getting-started-with-bot-sequences-and-segments) | One approach is not always adequate for automating a certain test scenario. Our drag and drop tool allows you to stitch together and re-use bot logic across tests. |

Reviewer 2: From a user perspective, this didn't inform enough about what the feature means to me.

```suggestion
| [Composability](core-concepts/bot-sequences/getting-started-with-bot-sequences-and-segments) | Define re-usable units of bot logic that perform a range of actions such as simulating inputs, invoking MonoBehaviours, and performing assertions against the game state. Then, stitch them together with our drag-and-drop tool to form different tests.  |
```

> | [Runtime, CI/CD, and Editor Compatibility](core-concepts/running-tests-in-the-unity-test-runner) | Our tools work in the Editor, in builds, and within CI/CD. Because of our integration approach, in most cases you will **not need to restart your game** to iterate on a test scenario. |

Reviewer 2: This hasn't been true for me. If the user is writing MonoBehaviours then they certainly will need to restart the game to iterate, and I anticipate that being the main use case at this stage of the product. I tried to add more detail to this statement but the stuff I tried out just reads to a user as "why are you even saying this if I don't know whether this will be true for me until I start using your SDK?"

The question after this suggestion is, what else can we write here? Because the name of the feature is now pretty much its entire description 🤔 

```suggestion
| [Runtime, CI/CD, and Editor Compatibility](core-concepts/running-tests-in-the-unity-test-runner) | Use our tools in the Editor, in builds, and within CI/CD. |
```

Reviewer: The only statement that is bolded in the whole table is **not need to restart your game**, and that draws attention. Is this really the most important sentence of the whole table? I personally don't think it is that important but I might be missing some context/user feedback

> | [Data Exploration and Validation](core-concepts/validation-suites) | Our dashboard allows developers to inspect and test the state of their game in a no-code fashion. |

Reviewer 2: Trying to think of a better word to use here than "Exploration" because a previous feature is "Exploratory Testing". 

```suggestion
| [Inspect and Validate Game State](core-concepts/validation-suites) | Review your recorded gameplay's screenshots, user inputs, and the internal state through our dashboard. Use these views to create test suites using no-code solutions. |
```

Reviewer: Can we say words about how we can show the whole play trace of an episode until the moment of the crash? That seems like a powerful feature

> ## Where do I start?

Reviewer 2: Always capitalize headers

```suggestion
## Where Do I Start?
```

The best place to start is with our [installation guide](getting-started/installing-regression-games). Once our SDK is added to your project, you can [create your first

> test automation](getting-started/creating-your-first-automated-test).

Reviewer 2: I interpret "test automation" as the practice, and an "automated test" as a concrete test case to run.

```suggestion
automated test](getting-started/creating-your-first-automated-test).
```

> If you are already familiar with the framework, check out these more advanced guides for test inspiration:

Reviewer 2: What framework? The SDK?

```suggestion
If you are already familiar with our framework, check out these more advanced guides:
```

* [Interacting with User Interfaces using Computer Vision](guides/interacting-with-ui-using-computer-vision)
* [Finding Game-Breaking Bugs with Stability Testing](guides/finding-game-breaking-bugs-with-stability-testing)
* [Writing Advanced Validations](guides/writing-advanced-validations)
* [Running Tests in Editor and in Builds](guides/running-tests-in-editor-and-in-builds)

> You can also check out our [existing game samples](additional-game-samples) or the [live demo of our validation tool](https://play.regression.gg/validations-demo).

Reviewer 2: I'd move this higher up, possibly even before the installation link. If someone is coming to this doc to figure out whether they want to give RG a try then showing them the live demo is high impact with minimal effort on their part. Not sure what the additional-game-samples page is supposed to contain at this point, so can't say for that link.

Author: Moved around! But also open to changing this a bit.

## Our Mission

Our mission is to transform the landscape of game development by tackling one of its most persistent challenges: automated testing.

Our tools and standards will make effective end-to-end automated testing of games a reality, in a way that actually saves you time and pain. We imagine a future where automated tests handle most functional testing scenarios, freeing up valuable time for developers to focus on the more challenging and creative aspects of game development. Regression Games is making this possible by combining old school algorithms, cutting-edge machine learning techniques, and a focus on the developer experience.

Our vision is to build tools with the community that actually help you be more productive. There are three core principles driving the development of this framework:
- **Ease of use** - Developers should be able to integrate and create new tests with minimal effort.
- **Reliability** - In the dynamic world of games, reliability is key. Our framework is designed to handle the unexpected and give you the tools to create robust, dependable tests.

> - **Data driven** - Our framework collects comprehensive data, empowering you to build smarter automations and create more effective tests.

Reviewer 3: I feel that `Data driven` is the least punchy of these principles. Trying to think of an alternative 🤔

Author: Open to suggestions!

sidebar_label: 'What is Regression Games?'
---

# Regression Games

> Regression Games is making it easy to create automated tests for your game.

Reviewer 2: ```suggestion
Regression Games makes it easy to create automated tests for your game.
```

## Highlighted Features

| Feature | Description |
| --- | --- |
| [Plug-and-Play Setup](getting-started/installing-regression-games) | Simply import our SDK package and add one GameObject to your scene to instantly begin using our tools. Custom integration code is rarely needed to get started. |

> | [Get Started Quickly with Recording and Input Playback](getting-started/creating-your-first-automated-test) | Create functional tests by recording a play-through of your scenario and preforming validations against automated playback. |

Reviewer 2: ```suggestion
| [Get Started Quickly with Recording and Input Playback](getting-started/creating-your-first-automated-test) | Create functional tests by recording a play-through of your scenario and performing validations against automated playback. |
```

| [Computer Vision Capabilities](guides/interacting-with-ui-using-computer-vision) | In some cases, a bot needs to _see_ your game to understand it. Our CV capabilities, such as OCR and object detection, can extract information from in-game screenshots to help drive behaviors and validate game state during runtime. |
| [Exploratory Testing](guides/finding-game-breaking-bugs-with-stability-testing) | Run more than just predefined scenarios - tools such as our Monkey Bot can dynamically explore your game to find edge cases. |
| [Composability](core-concepts/bot-sequences/getting-started-with-bot-sequences-and-segments) | Define re-usable units of bot logic that perform a range of actions such as simulating inputs, invoking MonoBehaviours, and performing assertions against the game state. Then, stitch them together with our drag-and-drop tool to form different tests.  |
| [Runtime, CI/CD, and Editor Compatibility](core-concepts/running-tests-in-the-unity-test-runner) | Use our tools in the Editor, in builds, and within CI/CD. |

> | [Inspect and Validate Game State](core-concepts/validation-suites) | Review your recorded gameplay's screenshots, user inputs, and the internal state through our dashboard. Use these views to create test suites using no-code solutions. |

Reviewer 2: ```suggestion
| [Inspect and Validate Game State](core-concepts/validation-suites) | Review your recorded gameplay's screenshots, user inputs, and internal game state through our dashboard. Use these views to create test suites using no-code solutions. |
```

## Where Do I Start?

> Before installing our SDK, you can always try our [live demo](https://play.regression.gg/validations-demo). Once you are ready to hop in, follow the

Reviewer 2: ```suggestion
See our [live validations demo](https://play.regression.gg/validations-demo) for an peek into the tools our SDK can unlock for your game. Once you are ready to hop in, follow the
```

> [installation guide](getting-started/installing-regression-games), after which you can 

Reviewer 2: ```suggestion
[installation guide](getting-started/installing-regression-games) and
```

[create your first automated test](getting-started/creating-your-first-automated-test).

If you are already familiar with our framework, check out these more advanced guides:

* [Interacting with User Interfaces using Computer Vision](guides/interacting-with-ui-using-computer-vision)
* [Finding Game-Breaking Bugs with Stability Testing](guides/finding-game-breaking-bugs-with-stability-testing)
* [Writing Advanced Validations](guides/writing-advanced-validations)
* [Running Tests in Editor and in Builds](guides/running-tests-in-editor-and-in-builds)

> You can also check out our [existing game samples](additional-game-samples).

Reviewer 2: what is the additional-game-samples page eventually going to contain / how will it help users? It'd be a good idea to throw that into this sentence ("You can check out our existing game samples for x")

Author: Made this a little more clear. Can update further as well once we add more samples!

sidebar_label: 'What is Regression Games?'
---

# Regression Games

> Regression Games is on a mission to transform the landscape of game development by tackling one of its most persistent challenges: automated testing.

Author: Done!

> We are developing the tools and standards that will make effective end-to-end automated testing of games a reality, in a way that actually saves you

Author: Done!

time and pain. We are imagining a future where automated tests can handle the most complex and repetitive scenarios, freeing up valuable time for developers to focus on
the most challenging and creative aspects of game development. By combining old school algorithms, cutting-edge machine learning techniques, and a focus
on the developer experience, Regression Games is building tools that make this end-to-end testing possible.

> - **Unleash Your Creativity:** By automating complex and repetitive testing scenarios, we free up your time to focus on what truly matters - creating amazing bug-free games.

Author: Added

> - **Cutting-Edge Technology:** We blend traditional algorithms with state-of-the-art machine learning, creating a powerful synergy that tackles the unique challenges of game testing.

Author: Added

> - **Developer-Centric Approach:** Our tools are built by developers, for developers. We understand your pain points and have crafted solutions that seamlessly integrate into your workflow.

Author: Removed

> ## What is our vision?

Reviewer 2: Not sold on this section being titled "vision". It reads more like the previous section is our high-level vision, and this one is more about our core principles for the implementation of that vision. I almost want to combine these sections because I don't see a strong distinction between the two. I'll give this another look once we clarify what these two sections is meant to convey to the reader (or brainstorm some edits with you over slack)

Author: Combined

> It's not secret that development teams are hesitant to adopt automated testing. If not done correctly, tests can become a burden rather than a way

Reviewer 2: ```suggestion
It's no secret that development teams are hesitant to adopt automated testing. If not done correctly, tests can become a burden rather than a way
```

to supercharge your development process.

> Our vision is to build tools with the community that actually help you be more productive. There are four core principles driving the development of this framework:

Reviewer 2: ```suggestion
Our vision is to build tools with the community that actually help you be more productive. There are three core principles driving the development of this framework:
```

Author: Uhhhhhhhhh fixed XD

- **Ease of use** - Developers should be able to integrate and create new tests with minimal effort.
- **Reliability** - In the dynamic world of games, reliability is key. Our framework is designed to handle the unexpected and give you the tools to create robust, dependable tests.
- **Data driven** - Our framework collects comprehensive data, empowering you to build smarter automations and create more effective tests.

Automated testing of games is challenging due to the complexity and dynamic behavior of games. This is why we
provide a variety of approaches to create **bots** that can handle the unique constraints of your game. We aren't locked into either old school
AI approaches or LLMs, but rather utilize any technique when it will actually result in a bot that can test your game reliably.

## Highlighted Features

> The table below highlights some of the main features of the framework.

Reviewer 2: Imo you don't need a primer statement if the entire section _is_ this table. It'll be clear to readers what the contents of the table are.

Author: Removed

| Feature | Description |
| --- | --- |

> | [Plug-and-Play Setup](getting-started/installing-regression-games) | Simply add our SDK and add one GameObject to your scene to begin instantly collecting detailed information for tests and bots. Custom integration code is rarely needed to get started. |

Reviewer 2: ```suggestion
| [Plug-and-Play Setup](getting-started/installing-regression-games) | Simply install our SDK and add one GameObject to your scene to instantly begin using our tools. Custom integration code is rarely needed to get started. |
```

Author: Added

> | [Get Started Quickly with Session Recording](getting-started/creating-your-first-automated-test) | Developers can create a new test simply by recording a play through of their scenario. |

Reviewer 2: We need to make a concerted effort to NOT conflate recording with gameplay sessions. I intentionally wrote the recording-gameplay and gameplay-session pages to make it clear that recording creates its own set of data which is _replicated_ in the gameplay-session, but that the session resources are used for entirely different tools (ie. validation suites) than the local resources are.

I also really don't like that we were referring to Gameplay Sessions as "tests" because they are used to create tests but the the recording isn't a test all by itself (especially after the discussions we've had about raw playback rarely being useful compared to running bot sequences)

```suggestion
| [Get Started Quickly with Recording and Input Playback](getting-started/creating-your-first-automated-test) | Developers can create functional tests by recording a play-through of their scenario and preforming validations against automated playback. |
```

Author: Fixed

> | [Computer Vision Capabilities](guides/interacting-with-ui-using-computer-vision) | In some cases, a bot needs to see the game to understand it. Our CV capabilities can extract information from game screenshots to help drive bot behavior and validate game state using approaches like OCR and object detection. |

Reviewer 2: ```suggestion
| [Computer Vision Capabilities](guides/interacting-with-ui-using-computer-vision) | In some cases, a bot needs to see the game to understand it. Our CV capabilities, such as OCR and object detection, can extract information from in-game screenshots to help drive behaviors and validate game state during runtime. |
```

Author: Edited

> | [Exploratory Testing](guides/finding-game-breaking-bugs-with-stability-testing) | The framework collects a wide variety of data, used in both testing and bot building. |

Reviewer 2: This title doesn't sound related to the description. Really, the description is a nothing-burger that basically says "yeah, we collect data" with no indication of what that data is or why the user should care

Author: Not sure how this got left over, I edited this to better reflect the title.

| [Composability](core-concepts/bot-sequences/getting-started-with-bot-sequences-and-segments) | One approach is not always adequate for automating a certain test scenario. Our drag and drop tool allows you to stitch together and re-use bot logic across tests. |
| [Runtime, CI/CD, and Editor Compatibility](core-concepts/running-tests-in-the-unity-test-runner) | Our tools work in the Editor, in builds, and within CI/CD. Because of our integration approach, in most cases you will **not need to restart your game** to iterate on a test scenario. |
| [Data Exploration and Validation](core-concepts/validation-suites) | Our dashboard allows developers to inspect and test the state of their game in a no-code fashion. |

## Where do I start?

> The best place to start is with our [installation guide](getting-started/installing-regression-games). Once the package is added, you can [create your first

Reviewer 2: ```suggestion
The best place to start is with our [installation guide](getting-started/installing-regression-games). Once our SDK is added to your project, you can [create your first
```

Author: Fixed.
